### ISSUE TITLE
Joined futures

### ISSUE THREAD

**Issue by kourge** [UrbanDoor/mobx-failable#4]
Joined futures allow us to combine multiple futures into one future.

## Motivation

Often we want to do something such as render data only when two different failables are both successes. One way we can already achieve this is by performing nested matches:

```ts
class EntryState {
  @observable user = new Failable<User>();
  @observable post = new Failable<Post>();
}

@observer
class Entry extends React.Component<{st: EntryState}> {
  render() {
    const {user: userF, post: postF} = this.props.st;
    return userF.match({
      success: user => (
        postF.match({
          success: post => (
            // use both `user` and `post`
          ),
          failure: e => null,
          pending: () => null,
        })
      ),
      failure: e => null,
      pending: () => null,
    });
  }
}
```

This excessively increases the number of cases to handle (the "pyramid of doom") as the number of successful things we want increases.

## Illustration

There is prior art in solving this problem in the JavaScript Promise API. [`Promise.all`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) allows us to combine multiple promises, essentially allowing us to "join" them. With joined futures, the above code becomes:

```ts
class EntryState {
  @observable user = new Failable<User>();
  @observable post = new Failable<Post>();
}

@observer
class Entry extends React.Component<{st: EntryState}> {
  render() {
    const {user: userF, post: postF} = this.props.st;
    return Failable.all({user: userF, post: postF}).match({
      success: ({user, post}) => (
        // use both `user` and `post
      ),
      failure: e => null,
      pending: () => null,
    });
  }
}
```

Note that in this example, we exploit JavaScript destructuring assignments and object literal shorthands to avoid boilerplate code.

## Interlude: `Promise.all`, prior art

In an earlier passage, `Promise.all` was mentioned. The signature of `Promise.all` is comprised of roughly the following definition:

```ts
interface PromiseConstructor {
  // higher-arity overloads, descending
  // ...
  all<T1, T2, T3>(values: [T1 | PromiseLike<T1>, T2 | PromiseLike<T2>, T3 | PromiseLike<T3>]): Promise<[T1, T2, T3]>;
  all<T1, T2>(values: [T1 | PromiseLike<T1>, T2 | PromiseLike<T2>]): Promise<[T1, T2]>;
  all<T>(values: (T | PromiseLike<T>)[]): Promise<T[]>;
}
```

- Each of the earlier overloads — containing multiple type parameters (e.g. `T1`, `T2`) as well as annotating `values` and the return type as a tuple type —  is considered _heterogeneous joining_, because each type parameter is a distinct type.
- The last overload with a single generic type — taking an array of promises and joining them into a single promise of an array — is called _homogenous joining_, because although multiple values are involved, only one distinct type is involved.

The explosion of high arity overloads for heterogenous joining illustrates how such an API is still hard for TypeScript's type system to describe, since it does not yet support [variadic kinds](https://github.com/Microsoft/TypeScript/issues/5453).

There is, however, hope. Heterogenous joining is usually a bounded set rather than unbounded. In other words, at compile time, the number of distinct types is usually known. Using two upcoming TypeScript features, [conditional types](https://github.com/Microsoft/TypeScript/pull/21316) and [type inference in conditional types](https://github.com/Microsoft/TypeScript/pull/21496), we can hypothetically rewrite `Promise.all`'s API to be shaped like an object:

```ts
type UnpackPromise<T> = T extends PromiseLike<infer U> ? U : T;

interface PromiseConstructor {
  all<T>(values: (T | PromiseLike<T>)[]): Promise<T[]>;
  all<T extends object>(values: T): Promise<{[K in keyof T]: UnpackPromise<T[K]>}>;
}
```

This hypothetical rewrite would allow us to write `Promise.all` like this:

```ts
Promise.all({
  user: fetchUser(userId),
  post: fetchPost(postId),
}).then(({user, post}) => {
  // use both `user` and `post`
});
```

## Specification

The proposed joined futures would be defined roughly as the following, with the `export` modifier omitted for brevity:

```ts
namespace Future {
  type Unpacked<T> = T extends Future<infer U> ? U : T;

  function all<T>(values: Future<T>[]): Future<T[]>;
  function all<T extends object>(values: T): Future<{[K in keyof T]: Unpacked<T[K]>}>;
}

namespace Failable {
  type Unpacked<T> = T extends Failable<infer U> ? U : T;

  function all<T>(values: Failable<T>[]): Failable<T[]>;
  function all<T extends object>(values: T): Failable<{[K in keyof T]: Unpacked<T[K]>}>;
}

namespace Loadable {
  type Unpacked<T> = T extends Loadable<infer U> ? U : T;

  function all<T>(values: Loadable<T>[]): Loadable<T[]>;
  function all<T extends object>(values: T): Loadable<{[K in keyof T]: Unpacked<T[K]>}>;
}
```

## Behavior

As with all monad-like structures with mapping behavior, we must be exact when specifying exactly _how_ different failables and loadables combine. For example, per the specification, `Promise.all` "rejects with the reason of the first passed promise that rejects". In other words, if one of the passed promises rejects, then the resulting combined promise rejects, even if all the other promises resolve.

### `Failable`

On an attempt to join multiple failables, the following decisions are made in the order they appear. Therefore earlier cases take precedence over latter cases.

When given an empty object or empty array, the joined failable is a success whose value is an empty array (giving the type `Failable<any[]>`) or empty object (giving the type `Failable<{}>`).

When any failable among multiple given failables is a pending, the joined failable is a pending.

When any failable among multiple given failables is a failure, the joined failable is a failure:

- If given an array, the error value of the failure is that of the first failable that is a failure.
- If given an object, the error value of the failure is arbitrary and random. (See later sections for ways around this limitation.)

Lastly, when all failables given are successes, the joined failable is a success:

- If given an array, the value of the success is that of an array containing the inner success values of every failable in the order that those failables appear in the array.
- If given an object, the value of the success is an object with the same keys as the given object, where every value is the inner success value of the failable corresponding to that key.

### `Loadable`

WIP

**Comment by jbraithwaite** [comment]
I'm working on implementing this.

**Comment by benbayard** [comment]
What is the expectation if there are multiple errors / failures?

**Comment by jbraithwaite** [comment]
I'm going with:

> * If given an array, the error value of the failure is that of the first failable that is a failure.
> * If given an object, the error value of the failure is arbitrary and random

**Comment by benbayard** [comment]
I don't think it should be arbitrary or random. IMO it should be an array of errors values, but the order is not guaranteed. I'd imagine the order is the order in which they failed. The problem here is that it's returning a slightly augmented `Failable` because `e` is `Error[]` vs `Error`. So the question is if we are okay extending Failable to include that definition

**Comment by kourge** [comment]
Another possibility is that it still has a single error, but the error is the last one that occurred.

I'm also open to the possibility of completely dropping the array version of `Failable.all` and only take objects. After all, an array is still an object, but one whose keys are numbers.

**Comment by benbayard** [comment]
Arrays are useful for tuples.

**Comment by benbayard** [comment]
Can we have our own error type which can have 1 message, or many messages? So:

```
class FailableError extends Error {
  constructor(private messages: string[])
  get message() {
    return this.messages[0]
  }
}
```

**Comment by kourge** [comment]
I thought about that, but `Failable` defines its failures as containing plain `Error` types, so the consumer of the API would be forced to cast from `Error` to `FailableError` in this scenario.

Hmm. Let `A extends Array`, which is any array or tuple type. What if we instead return `A & Supplements`, where `Supplements` refers to "Supplementary computed properties" in #4.
